File: packages/vela-sdk/src/vela_sdk/sources/registry.py
"""Global MCP-source-handler registry.

Module-level dict; lookups are O(1). Re-registering the same name raises
to surface duplicate-bootstrap bugs early. `clear_sources()` is exported
for tests only.
"""

from typing import Optional

from vela_sdk.sources.types import SourceHandler

_handlers: dict[str, SourceHandler] = {}


def register_source(name: str, handler: SourceHandler) -> None:
    """Register a handler for a source namespace (e.g. "devops").

    Raises:
        ValueError: if the name is already registered.
    """
    if name in _handlers:
        raise ValueError(f"source '{name}' already registered")
    _handlers[name] = handler


def resolve_source(name: str) -> Optional[SourceHandler]:
    """Look up a registered handler. Returns None if not registered."""
    return _handlers.get(name)


def clear_sources() -> None:
    """Test utility — wipes the registry so suites stay isolated."""
    _handlers.clear()

File: packages/vela-sdk-ts/src/sources/registry.ts
/**
 * Global MCP-source-handler registry.
 *
 * Module-level Map; lookups are O(1). Re-registering the same name throws
 * to surface duplicate-bootstrap bugs early. `clearSources()` is exported
 * for tests only.
 */

import type { SourceHandler } from "./types.js";

const handlers = new Map<string, SourceHandler>();

/**
 * Register a handler for a source namespace (e.g. "devops").
 *
 * @throws if the name is already registered.
 */
export function registerSource(name: string, handler: SourceHandler): void {
  if (handlers.has(name)) {
    throw new Error(`source '${name}' already registered`);
  }
  handlers.set(name, handler);
}

/** Look up a registered handler. Returns undefined if not registered. */
export function resolveSource(name: string): SourceHandler | undefined {
  return handlers.get(name);
}

/** Test utility — wipes the registry so suites stay isolated. */
export function clearSources(): void {
  handlers.clear();
}
